django/db/models/sql/compiler.py

```diff
@@ -345,7 +345,12 @@ def get_order_by(self):
                         resolved.set_source_expressions([RawSQL('%d' % (idx + 1), ())])
                         break
                 else:
-                    raise DatabaseError('ORDER BY term does not match any column in the result set.')
+                    if col_alias:
+                        raise DatabaseError('ORDER BY term does not match any column in the result set.')
+                    # Add column used in ORDER BY clause without an alias to
+                    # selected columns.
+                    self.query.add_select_col(src)
```

Comment:
Composed queryset can be used multiple times, that's why adding columns is not a proper solution. We could use instead:
```python
self.query.set_values([*self.query.values_select, src.output_field.name])
```
but we set list of values in combined queries only once (see [compiler.py#L428-L433](https://github.com/django/django/blob/6e8303d49ba55e7393c7ef5553b240ae0d2f0c00/django/db/models/sql/compiler.py#L428-L433)), so e.g. in `test_union_with_values_list_and_order()` the first subtest sets the list of columns to `('order', 'id')` which remain the same for all subtests (that's why the last two fail). I added a separate ticket that should be fixed in advance ([29834](https://code.djangoproject.com/ticket/29834)).

Reply:
thanks @felixxm, I'll check check asap.

Reply:
I've submitted patch https://github.com/django/django/pull/11490.

Reply:
OK, so after #11490 `add_select_col()` works properly because `self.query` is created by a compiler so it is not reuse and doesn't mutate combined queries.